examples/control_character/run.py

```diff
@@ -0,0 +1,193 @@
+#
+# Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+#
+
+# Before running this file, rename user_config.yaml.example -> user_config.yaml and modify it with appropriate paths for your system.
+
+import argparse
+import os
+import numpy as np
+import pandas as pd
+import spear
+
+
+def get_data_frame(skeletal_mesh_transforms):
+    columns = np.array([
+        name + "." + field + "." + axis
+        for name, transform in skeletal_mesh_transforms.items()
+        for field, vector in transform.items()
+        for axis, val in vector.items()
+    ]).ravel()
+    data = np.array([
+        value
+        for name, transform in skeletal_mesh_transforms.items()
+        for field, vector in transform.items()
+        for axis, value in vector.items()
+    ]).reshape(1, -1)
+    return pd.DataFrame(columns=columns, data=data)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--actions_file", default=os.path.realpath(os.path.join(os.path.dirname(__file__), "actions.csv")))
+    args = parser.parse_args()
+
+    np.set_printoptions(linewidth=200)
+
+    # load config
+    config = spear.get_config(user_config_files=[os.path.realpath(os.path.join(os.path.dirname(__file__), "user_config.yaml"))])
+    spear.configure_system(config)
+    instance = spear.Instance(config)
+
+    with instance.begin_frame():
+        gameplay_statics_uclass = instance.unreal_service.load_class(class_name="UObject", outer=0, name="/Script/Engine.GameplayStatics")
```

Comment:
Storing per-character state in an array will clean up the code in multiple places.

```
        gameplay_statics_static_class = instance.unreal_service.get_static_class(class_name="UGameplayStatics")
        character_movement_component_static_class = instance.unreal_service.get_static_class(class_name="UCharacterMovementComponent")
        skeletal_mesh_component_static_class = instance.unreal_service.get_static_class(class_name="USkeletalMeshComponent")

        gameplay_statics_default_object = instance.unreal_service.get_default_object(uclass=gameplay_statics_uclass, create_if_needed=False)

        #
        # initialize characters
        #

        character_descs = [
            {"name": "character_0", "location": {"X": 0.0, "Y": 100.0, "Z": 150.0}, "rotation": {"Roll": 0.0, "Pitch": 0.0, "Yaw": 0.0}},
            {"name": "character_1", "location": {"X": 0.0, "Y": 200.0, "Z": 150.0}, "rotation": {"Roll": 0.0, "Pitch": 0.0, "Yaw": 0.0}}]

        bp_character_uclass = instance.unreal_service.load_object(
            class_name="UClass", outer=0, name="/Game/ThirdPerson/Blueprints/BP_ThirdPersonCharacter.BP_ThirdPersonCharacter_C")

        characters = []
        for character_desc in character_descs:

            character = {}
            character["actor"] = instance.unreal_service.spawn_actor_from_uclass(
                uclass=bp_character_uclass,
                location=character_desc["location"],
                rotation=character_desc["rotation"],
                spawn_parameters={"Name": character_desc["name"], "SpawnCollisionHandlingOverride": "AlwaysSpawn"})
            character["movement_component"] = instance.unreal_service.get_component_by_name(actor=character["actor"], component_name="...")
            character["skeletal_mesh_component"] = instance.unreal_service.get_component_by_name(actor=character["actor"], component_name="...")

            # get bone names
            return_values = instance.unreal_service.call_function(uobject=character["skeletal_mesh_component"], ufunction=get_num_bones_ufunc)
            num_bones = return_values["ReturnValue"]
            character["bone_names"] = []
            for i in range(num_bones):
                return_values = instance.unreal_service.call_function(
                    uobject=character["skeletal_mesh_component"], ufunction=get_bone_name_ufunc, args={"BoneIndex": i})
                bone_name = return_values["ReturnValue"]
                character["bone_names"].append(bone_name)

            # configure character to be controlled without possessing
            instance.unreal_service.set_object_properties_for_uobject(uobject=character["movement_component"], properties={"bRunPhysicsWithNoController": True})
            instance.unreal_service.call_function(uobject=character["movement_component"], ufunction=set_movement_mode_ufunc, args={"NewMovementMode": "MOVE_Walking"})
            instance.input_service.setup_player_input_component(actor=character["actor"])

            characters.append(character)

        assert characters[0]["bone_names"] == characters[1]["bone_names"]
        bone_names = characters[0]["bone_names"]

        # update the static mesh for character_0
        manny_simple_skeletal_mesh = instance.unreal_service.load_object(
            class_name="UObject", outer=0, name="/Game/Characters/Mannequins/Meshes/SKM_Manny_Simple.SKM_Manny_Simple")
```